### ISSUE TITLE
Build error related to std::type_info

### ISSUE THREAD

**Issue by NorthSailor** [gnustep/libobjc2#69]
The following error appears when building `libobjc2` on Ubuntu 18.04 with Clang 6.0.0.

```
Scanning dependencies of target objc
[  0%] Building C object CMakeFiles/objc.dir/abi_version.c.o
[  1%] Building C object CMakeFiles/objc.dir/alias_table.c.o
[  1%] Building C object CMakeFiles/objc.dir/block_to_imp.c.o
[  2%] Building C object CMakeFiles/objc.dir/caps.c.o
[  2%] Building C object CMakeFiles/objc.dir/category_loader.c.o
[  3%] Building C object CMakeFiles/objc.dir/class_table.c.o
[  4%] Building C object CMakeFiles/objc.dir/dtable.c.o
[  4%] Building C object CMakeFiles/objc.dir/eh_personality.c.o
[  5%] Building C object CMakeFiles/objc.dir/encoding2.c.o
[  5%] Building C object CMakeFiles/objc.dir/hash_table.c.o
[  6%] Building C object CMakeFiles/objc.dir/hooks.c.o
[  6%] Building C object CMakeFiles/objc.dir/ivar.c.o
[  7%] Building C object CMakeFiles/objc.dir/legacy_malloc.c.o
[  7%] Building C object CMakeFiles/objc.dir/loader.c.o
[  8%] Building C object CMakeFiles/objc.dir/mutation.m.o
[  8%] Building C object CMakeFiles/objc.dir/protocol.c.o
[  9%] Building C object CMakeFiles/objc.dir/runtime.c.o
[ 10%] Building C object CMakeFiles/objc.dir/sarray2.c.o
[ 10%] Building C object CMakeFiles/objc.dir/selector_table.c.o
[ 11%] Building C object CMakeFiles/objc.dir/sendmsg2.c.o
[ 11%] Building C object CMakeFiles/objc.dir/statics_loader.c.o
[ 12%] Building C object CMakeFiles/objc.dir/toydispatch.c.o
[ 12%] Building ASM object CMakeFiles/objc.dir/block_trampolines.S.o
[ 13%] Building ASM object CMakeFiles/objc.dir/objc_msgSend.S.o
[ 13%] Building C object CMakeFiles/objc.dir/NSBlocks.m.o
[ 14%] Building C object CMakeFiles/objc.dir/Protocol2.m.o
[ 15%] Building C object CMakeFiles/objc.dir/arc.m.o
[ 15%] Building C object CMakeFiles/objc.dir/associate.m.o
[ 16%] Building C object CMakeFiles/objc.dir/blocks_runtime.m.o
[ 16%] Building C object CMakeFiles/objc.dir/properties.m.o
[ 17%] Building C object CMakeFiles/objc.dir/gc_none.c.o
[ 17%] Building CXX object CMakeFiles/objc.dir/objcxx_eh.cc.o
/home/nsailor/Projects/GNUstep/libobjc2/objcxx_eh.cc:25:8: error: redefinition of 'type_info'
        class type_info
              ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/7.3.0/../../../../include/c++/7.3.0/typeinfo:88:9: note: previous definition is here
  class type_info
        ^
1 error generated.
CMakeFiles/objc.dir/build.make:790: recipe for target 'CMakeFiles/objc.dir/objcxx_eh.cc.o' failed
make[2]: *** [CMakeFiles/objc.dir/objcxx_eh.cc.o] Error 1
CMakeFiles/Makefile2:99: recipe for target 'CMakeFiles/objc.dir/all' failed
make[1]: *** [CMakeFiles/objc.dir/all] Error 2
Makefile:162: recipe for target 'all' failed
make: *** [all] Error 2
```

**Comment by vincentwoo** [comment]
I have this issue as well.

**Comment by NorthSailor** [comment]
@vincentwoo You can check my suggested patch at https://github.com/gnustep/libobjc2/pull/68

**Comment by eklitzke** [comment]
I think the source of confusion here is that if you build using Clang and don't have the LLVM libc++ headers installed, clang implicitly uses the GCC libstdc++ headers. While this *should* work in theory, it seems like a kind of fragile build configuration.

In addition to your patch at #68, what about an explicit check to see which C++ header implementation you're using? Cmake already prints a warning message if you try to build with GCC, so my proposed change would be to also warn when using the GCC libstd++ headers.

**Comment by davidchisnall** [comment]
Note: There is nothing wrong with using clang to compile code that links against libstdc++ and this is what people *should* do if libstdc++ is the target platform's default C++ standard library and is (intended to be supported) by libobjc2.

**Comment by dlech** [comment]
Has this actually been fixed? It seems that #68 only fixed this error:

```
/usr/bin/../lib/gcc/x86_64-linux-gnu/7.3.0/../../../../include/c++/7.3.0/bits/cxxabi_init_exception.h:63:7: error: exception specification in declaration does not match
      previous declaration
      __cxa_allocate_exception(size_t) _GLIBCXX_NOTHROW;
      ^
/home/nsailor/Projects/GNUstep/libobjc2/objcxx_eh.h:10:7: note: previous declaration is here
void *__cxa_allocate_exception(size_t thrown_size);
      ^
```

but not this error:

```
/home/nsailor/Projects/GNUstep/libobjc2/objcxx_eh.cc:26:8: error: redefinition of 'type_info'
        class type_info
              ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/7.3.0/../../../../include/c++/7.3.0/typeinfo:88:9: note: previous definition is here
  class type_info
```
